iptables.go

```diff
@@ -0,0 +1,208 @@
+		return fmt.Errorf("iptables init failed: %s", err.Error())
+	}
+
+	return nil
+}
+
+func (ipt *iptables) AddBan(ban types.BanApplication) error {
+	done := false
+
+	//we now have to know if ba is for an ipv4 or ipv6
+	//the obvious way would be to get the len of net.ParseIp(ba) but this is 16 internally even for ipv4.
+	//so we steal the ugly hack from https://github.com/asaskevich/govalidator/blob/3b2665001c4c24e3b076d1ca8c428049ecbb925b/validator.go#L501
+	if strings.Contains(ban.IpText, ":") {
+		if err := ipt.v6.addBan(ban); err != nil {
+			return fmt.Errorf("failed inserting ban ip '%s' for iptables ipv4 rule", ban.IpText)
+		}
+		done = true
+	}
+	if strings.Contains(ban.IpText, ".") {
+		if err := ipt.v4.addBan(ban); err != nil {
+			return fmt.Errorf("failed inserting ban ip '%s' for iptables ipv6 rule", ban.IpText)
+		}
+		done = true
+	}
+
+	if !done {
+		return fmt.Errorf("failed inserting ban: ip %s was not recognised", ban.IpText)
+	}
+
+	return nil
+}
+
+func (ipt *iptables) ShutDown() error {
+	err := ipt.v4.shutDown()
+	if err != nil {
+		return fmt.Errorf("iptables for ipv4 shutdown failed: %s", err.Error())
+	}
+	err = ipt.v6.shutDown()
+	if err != nil {
+		return fmt.Errorf("iptables for ipv6 shutdown failed: %s", err.Error())
+	}
+	return nil
+}
+
+func (ipt *iptables) DeleteBan(ban types.BanApplication) error {
+	done := false
+	if strings.Contains(ban.IpText, ":") {
+		if err := ipt.v6.deleteBan(ban); err != nil {
+			return fmt.Errorf("failed deleting ban")
+		}
+		done = true
+	}
+	if strings.Contains(ban.IpText, ".") {
+		if err := ipt.v4.deleteBan(ban); err != nil {
+			return fmt.Errorf("failed deleting ban")
+		}
+		done = true
+	}
+	if !done {
+		return fmt.Errorf("failed inserting ban: ip %s was not recognised", ban.IpText)
+	}
+	return nil
+}
+
+func (ipt *iptables) Run(dbCTX *sqlite.Context, frequency time.Duration) error {
+
+	lastTS := time.Now()
+	/*start by getting valid bans in db ^^ */
+	log.Infof("fetching existing bans from DB")
+	bansToAdd, err := getNewBan(dbCTX)
+	if err != nil {
+		return err
+	}
+	log.Infof("found %d bans in DB", len(bansToAdd))
+	for idx, ba := range bansToAdd {
+		log.Debugf("ban %d/%d", idx, len(bansToAdd))
+		if err := ipt.AddBan(ba); err != nil {
+			return err
+		}
+
+	}
+	/*go for loop*/
+	for {
+		// check if ipset set and iptables rules are still present. if not creat them
+		if err := ipt.v4.CheckAndCreate(); err != nil {
+			return err
+		}
+		if err := ipt.v6.CheckAndCreate(); err != nil {
+			return err
+		}
+		time.Sleep(frequency)
+
+		bas, err := getDeletedBan(dbCTX, lastTS)
+		if err != nil {
+			return err
+		}
+		if len(bas) > 0 {
+			log.Infof("%d bans to flush since %s", len(bas), lastTS)
+		}
+		for idx, ba := range bas {
+			log.Debugf("delete ban %d/%d", idx, len(bas))
+			if err := ipt.DeleteBan(ba); err != nil {
+				return err
+			}
+		}
+
+		bansToAdd, err := getNewBan(dbCTX)
+		if err != nil {
+			return err
+		}
+		fmt.Printf("Adding ban")
```

Comment:
maybe display the numbers of ban that are going to be added, ie `Adding %d bans`

Reply:
this is a debug, it will be removed ^^